### PR TITLE
fix(firestore, web): blob was being parsed as geopoint.

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

Separated the encoding and decoding because I wanted to import classes from different files with name clashes. The relevant fix is [here](https://github.com/FirebaseExtended/flutterfire/pull/6728/files#diff-ad2e8b21ea951364151827b99cea0ce04dbf437c8f696f418e9b6d6bf916970dR33) which allows for checking if the value is a `GeoPoint` class.  

## Related Issues

fixes https://github.com/FirebaseExtended/flutterfire/issues/3865

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore_web/lib/src/utils/decode_utility.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/// Class containing static utility methods to decode firestore data.
+class DecodeUtility {
+  /// Decodes the values on an incoming Map to their proper types.
+  static Map<String, dynamic>? decodeMapData(Map<String, dynamic>? data) {
+    if (data == null) {
+      return null;
+    }
+    Map<String, dynamic> output = Map.from(data);
+    output.updateAll((key, value) => valueDecode(value));
+    return output;
```
Comment (lesnitsky) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/utils/decode_utility.dart:22]:
```suggestion
    return data..updateAll((key, value) => valueDecode(value));
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/utils/decode_utility.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  /// Decodes the values on an incoming Array to their proper types.
+  static List<dynamic>? decodeArrayData(List<dynamic>? data) {
+    if (data == null) {
+      return null;
+    }
+    return List.from(data).map(valueDecode).toList();
```
Comment (lesnitsky) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/utils/decode_utility.dart:30]:
```suggestion
    return data.map(valueDecode).toList();
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/utils/utils.dart [hunk 3]
```diff
@@ -24,7 +24,7 @@ dynamic dartify(Object? jsObject) {
           (object! as TimestampJsImpl).toMillis());
     }
     if (util.instanceof(object, BlobConstructor)) {
-      return object! as Blob;
+      return object! as BlobJsImpl;
```
Comment (lesnitsky) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/utils/utils.dart:27]: you can avoid bang operator with early null check return

**Review by lesnitsky — CHANGES_REQUESTED** [review]

**Review by lesnitsky — APPROVED** [review]
LGTM

**Review by Ehesp — APPROVED** [review]

**Pull request merged**